docs/design.md

```diff
@@ -0,0 +1,279 @@
+# Indexer Reference Provider Design
+The goal of this reference implementation is to flesh out all the protocol details
+for data providers to be able to interact with indexer nodes. 
+
+## Overview
+- Data providers keep a local index of all the data they are providing.
+- Data providers are responsible for the advertisement of updates related to the data they are providing.
+        - Data providers must be able to uniquely identify the batches of data they are providing, and/or the
+        announcements they broadcast to the network in order to be able to serve indexer-node data ingestion requests.
+- Indexer nodes listen to advertisement from data providers. When they come across a previously unseeend
+advertisement for a data provider, they trigger a new ingestion session to sync with the provider. 
+
+## Data Structures
+Data providers and indexer nodes can use any internal representation they want for indexed data. They only
+data structures enforced by the protocol are in the interaction between the two.
+
+### Advertisements
+Advertisements are used by data providers to share updates about the data they are providing. Ideally,
+these advertisements are published in a shared pub-sub channel between providers and indexer nodes. However,
+some indexer nodes may choose to provide a dedicated endpoint to allow certain authenticated providers to push announcements and updates directly to them.
+- `Pull` Advertisements: Indexer-nodes pull advertisement from a shared pub-sub channel.
+- `Push` Advertisements: Authenticated providers can push advertisements to indexer-nodes.
```

Comment:
Does a miner have to be discovered first, or is the push supposed to be another mechanism for provider discovery?

Reply:
My reasoning here was that some indexers may be open to: 
- Accept new providers that proactively push advertisements to them (after the corresponding verification); 
- Accept advertisement pushes from already verified providers (because they are a priority?) so it these indexers don't have to wait to catch the advertisement from the pubsub channel.
It would be up to the indexer to config the policies it supports.

Reply:
OK, then somewhere in the doc state that a `Push` can also serve as a discovery mechanism.